### ISSUE TITLE
Auto-kill overlay registered tasks in TestBase.initialize

### ISSUE THREAD

**Issue by qstokkink** [Tribler/py-ipv8#970]
Many overlays will register periodical cleanup/maintenance tasks in their `__init__()`.  The pattern is as follows:

```python
class MyOverlay(Community):

    def __init__( ... ):
        ... setup stuff ...
        self.register_task(self.some_method_that_runs_forever)
```

The official IPv8 policy on this is: "don't do this". The reason is that these tasks can be scheduled before IPv8 is done loading all overlay instances and this makes your overlay harder to test (for example `deliver_messages` has very little means to check if a task is an infinite loop or not). These tasks should be defined using the `on_start` (`configuration.py`) or `finalize()` (`loader.py`) callbacks or use `DiscoveryStrategy` if the task performs network traffic and loops indefinitely.

Even though I still consider registering tasks in the `__init__` a bad practice, we can at least make this easier to test. In 99% of the cases these tasks consist of cleanup or refresh mechanisms. This means that, by default, `TestBase` can kill of all the tasks registered in the overlay's `__init__` in `TestBase.initialize` without disrupting any overlay logic. Everything will just magically work as expected.

If we implement a new method in `TestBase` to perform this task killing (essentially just `overlay.cancel_all_pending_tasks()`) and make the task killing itself an argument to `initialize` we can support both partial/custom task killing and disabling task killing altogether per `TestCase`. This deals with the 1% of cases where auto-killing causes problems.

### Examples

```python
class MyOverlay(Community):

    def __init__(self, *args, **kwargs):
        super().__init__(self, *args, **kwargs)

        self.register_task("does not bleed", self.important_task, interval=3.0)
        self.register_task("can be killed", self.unimportant_task, interval=300.0)

    def important_task(self):
        print("Good thing we did this, otherwise something would've crashed!")

    def unimportant_task(self):
        print("Run this once in a while to make sure everything is clean.")


class TestNormal(TestBase):

    def test_something(self):
        self.initialize(MyOverlay, 2)

        # Periodically invoked methods can be called manually to make tests deterministic and clean.
        self.overlay(0).important_task()

        self.overlay(0).something()

        # Nothing is running in the background, everything magically works.
        self.deliver_messages()


class TestCustom(TestBase):

    def kill_tasks(self, overlay):
        """
        Overwrite task killing with some custom logic.
        Only kill what can be killed.
        """
        overlay.cancel_pending_task("can be killed")

    def test_something(self):
        self.initialize(MyOverlay, 2)

        self.overlay(0).something()

        # Doing custom things comes with great responsibility.
        # The following will still fail to return before the timeout if `important_task` loops infinitely.
        # Even worse, if the tests are running on a potato this may invoke `important_task`
        # a differing amount of times as compared to a fast computer, if it loops indefinitely.
        self.deliver_messages()


class TestProfessional(TestBase):

    def test_something(self):
        self.initialize(MyOverlay, 2, kill_init_tasks=False)

        # Old behavior. Apparently the writer of the tests knows what they are doing.
        # You're relegated to `sleep` and awaiting custom `Future` instances in your overlay.
```

### Future work

At some point we should provide documentation on why and when you should use `DiscoveryStrategy` and `register_task` and where to spawn tasks.

**Comment by qstokkink** [comment]
**tl;dr:** calling `overlay.cancel_all_pending_tasks()` for all overlays created in `TestBase.initialize` is probably a good idea.

**Comment by qstokkink** [comment]
After implementing this I changed my mind and I will drop this idea.

A `TestCase` should observe functionality, not seek to define it: the class-under-test should be tested as-is. 

There is (1) no performance increase and (2) this feature introduces additional complexity, which is (3) also partially hidden.
